### DISCUSSION TITLE
Slow Pandas to_sql with mssql+pyodbc

### DISCUSSION THREAD

**Discussion by kmatt** [sqlalchemy/sqlalchemy#9436] [Usage Questions]
### Describe the bug

Compared to SQLAlchemy==1.4.46, writing a Pandas dataframe with `pandas.to_sql` using an SQLAlchemy 2.0.4 engine takes about 10X longer on average.

The same is observed creating the engine as

    engine = create_engine(uri, fast_executemany=True)

or

    engine = create_engine(uri, use_insertmanyvalues=True)

Since the data is written without exceptions from either SQLAlchemy or Pandas, what else could be used to determine the cause of the slow down?

Pandas chunksize has no measurable effect.

### SQLAlchemy Version in Use

2.0.4

### DBAPI (i.e. the database driver)

puodbc

### Database Vendor and Major Version

Azure SQL Database

### Python Version

3.10.6

### Operating system

Ubuntu 20.04

### To Reproduce

```python
import pandas as pd
import sqlalchemy as sa

uri =  mssql+pyodbc://***:***@***.database.windows.net/testdb?driver=ODBC+DRIVER+18+for+SQL+Server
engine = sa.create_engine(uri)

df = pd.read_parquet(filename)
df.to_sql(tablename, engine, if_exists='append', index=False)
```

### Error

No errors - only observed runtime differences

**Comment by zzzeek** [comment]
hi -

there's no reproduction case here so no evidence of a bug, we can advise you on measuring performance.

you want to start using `echo=True` on your create_engine() and observe if the SQL being emitted is different.  in particular, look for things like lots of COMMIT happening in case pandas is not using transactions correctly, since that's what this sounds like.

a full step by step for profiling is in the FAQ at https://docs.sqlalchemy.org/en/20/faq/performance.html#how-can-i-profile-a-sqlalchemy-powered-application - run through those steps and the issue can be identified.

**Reply by zzzeek** [reply]
try setting use_setinputsizes=False

    e = create_engine("...",  use_setinputsizes=False)

**Reply by kmatt** [reply]
`use_setinputsizes` does not appear to change this situation. See below.

**Comment by kmatt** [comment]
`sa.create_engine(connection_url, use_insertmanyvalues=True, insertmanyvalues_page_size=1000, echo=True)` shows a single INSERT statement for each row.

Reverting to `sa.create_engine(uri, fast_executemany=True, use_insertmanyvalues=False, echo=True)` shows a parameterized INSERT statement with a batch of rows being sent to it, and thus much shorter runtime.

I was reading https://docs.sqlalchemy.org/en/20/dialects/mssql.html#fast-executemany-mode to mean `use_insertmanyvalues` is preferred over PyODBC `fast_executemanyvalues` but in my case the latter appears to send far fewer transactions. Am I misinterpreting the documentation for 2.0?

**Reply by zzzeek** [reply]
> sa.create_engine(connection_url, use_insertmanyvalues=True, insertmanyvalues_page_size=1000, echo=True) shows a single INSERT statement for each row.

that's the wrong output.   there should be a single INSERT statement for every 1000 rows.  Would need a reproduction case.

> Am I misinterpreting the documentation for 2.0?

Yes, pretty much.   pyodbc fast_executemany does the same thing that SQLAlchemy's insertmanyvalues does, it's just in the logging you don't see if because that's SQLAlchemy's logging before data is sent to the driver.

also none of this has anything to do with "transactions", transactional boundaries are not changed. 

in our own testing, "insertmanyvalues" shows an order of magnitude performance increase for SQL Server since 1.4, see https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#benchmarks  

however that requires that the feature is being enabled which it seems pandas is somehow breaking.

**Reply by zzzeek** [reply]
there seem to be many ways they build and execute INSERT statements:

https://github.com/pandas-dev/pandas/blob/main/pandas/io/sql.py

it's not clear which one is taking effect in your case.    

Here's a demo for SQL Server, try running this

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base

Base = declarative_base()


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    data = Column(String)


e = create_engine(
    "mssql+pyodbc://scott:tiger^5HHH@mssql2017:1433/"
    "test?driver=ODBC+Driver+17+for+SQL+Server",
    echo=True,
)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)

with e.begin() as conn:
    conn.execute(
        A.__table__.insert(),
        [{"id": i, "data": f"data {i}"} for i in range(10000)],
    )
```



logging output is a handful of INSERT statements that look like this:

```sql
2023-03-07 13:42:32,416 INFO sqlalchemy.engine.Engine INSERT INTO a (id, data) VALUES (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (? ... 7680 characters truncated ... , ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?)
2023-03-07 13:42:32,416 INFO sqlalchemy.engine.Engine [insertmanyvalues batch 10 of 10] (9000, 'data 9000', 9001, 'data 9001', 9002, 'data 9002', 9003, 'data 9003', 9004, 'data 9004', 9005, 'data 9005', 9006, 'data 9006', 9007, 'data 9007', 9008, 'data 9008', 9009, 'data 9009', 9010, 'data 9010', 9011, 'data 9011', 9012, 'data 9012', 9013, 'data 9013', 9014, 'data 9014', 9015, 'data 9015', 9016, 'data 9016', 9017, 'data 9017', 9018, 'data 9018', 9019, 'data 9019', 9020, 'data 9020', 9021, 'data 9021', 9022, 'data 9022', 9023, 'data 9023', 9024, 'data 9024' ... 1900 parameters truncated ... 9975, 'data 9975', 9976, 'data 9976', 9977, 'data 9977', 9978, 'data 9978', 9979, 'data 9979', 9980, 'data 9980', 9981, 'data 9981', 9982, 'data 9982', 9983, 'data 9983', 9984, 'data 9984', 9985, 'data 9985', 9986, 'data 9986', 9987, 'data 9987', 9988, 'data 9988', 9989, 'data 9989', 9990, 'data 9990', 9991, 'data 9991', 9992, 'data 9992', 9993, 'data 9993', 9994, 'data 9994', 9995, 'data 9995', 9996, 'data 9996', 9997, 'data 9997', 9998, 'data 9998', 9999, 'data 9999')
```

that's what you should see.

**Reply by zzzeek** [reply]
I just ran this script:

```py
import pandas as pd
import sqlalchemy as sa

engine = sa.create_engine(
    "mssql+pyodbc://scott:tiger^5HHH@mssql2017:1433/"
    "test?driver=ODBC+Driver+17+for+SQL+Server",
    echo=True,
)


df = pd.read_csv("/home/classic/pandas_file.csv")
df.to_sql("my_table", engine, if_exists="append", index=False)

```

using this file:  https://github.com/numpy/numpy/blob/main/numpy/core/tests/data/umath-validation-set-tanh.csv

statements look like:

```
 INSERT INTO my_table (dtype, input, output, ulperrortol) VALUES (?, ?, ?, ?), (?, ?, ?, ?), (?, ?, ?, ?), (?, ?, ?, ?), (?, ?, ?, ?), (?, ?, ?, ?), (?, ?, ?, ?), (?, ?, ?, ?), (?, ?, ?, ?), (?, ?, ?, ?), (?, ?, ?, ?), (?, ?, ?, ?), (?, ?, ?, ?), (?,  ... 5032 characters truncated ... ?), (?, ?, ?, ?), (?, ?, ?, ?), (?, ?, ?, ?), (?, ?, ?, ?), (?, ?, ?, ?), (?, ?, ?, ?), (?, ?, ?, ?)
2023-03-07 13:56:14,309 INFO sqlalchemy.engine.Engine [insertmanyvalues batch 3 of 3] ('np.float64', '0x3fd80238f6b00470', '0x3fd6f143be1617d6', 2, 'np.float64', '0xbfd05bfeb3a0b7fe', '0xbfd0031ab3455e15', 2, 'np.float64', '0xffc3a50351274a08', '0xbff0000000000000', 2, 'np.float64', '0xffd8f4241cb1e848', '0xbff0000000000000', 2, 'np.float64', '0xbfca72a88c34e550', '0xbfca13ebe85f2aca', 2, 'np.float64', '0x3fd47d683ba8fad0', '0x3fd3d13f1176ed8c', 2, 'np.float64', '0x3fb6418e642c831d', '0x3fb6333ebe479ff2', 2, 'np.float64', '0x800fde8e023fbd1c', '0x800fde8e023fbd1c', 2, 'np.float64', '0x8001fb01e323f605', '0x8001fb01e323f605', 2, 'np.float64', '0x3febb21ff9f76440', '0x3fe65ed788d52fee', 2, 'np.float64', '0x3fe47553ffe8eaa8', '0x3fe20fe01f853603', 2, 'np.float64', '0x7fca20b3f9344167', '0x3ff0000000000000', 2, 'np.float64', '0x3fe704f4ec6e09ea' ... 1420 parameters truncated ... '0x3ff0000000000000', 2, 'np.float64', '0x3fe93debccf27bd8', '0x3fe50c27727917f0', 2, 'np.float64', '0xe5ead05bcbd5a', '0xe5ead05bcbd5a', 2, 'np.float64', '0xbfebbbc4cff7778a', '0xbfe663c4ca003bbf', 2, 'np.float64', '0xbfea343eb474687e', '0xbfe59529f73ea151', 2, 'np.float64', '0x3fbe74a5963ce94b', '0x3fbe50123ed05d8d', 2, 'np.float64', '0x3fd31d3a5d263a75', '0x3fd290c026cb38a5', 2, 'np.float64', '0xbfd79908acaf3212', '0xbfd695620e31c3c6', 2, 'np.float64', '0xbfc26a350324d46c', '0xbfc249f335f3e465', 2, 'np.float64', '0xbfac38d5583871b0', '0xbfac31866d12a45e', 2, 'np.float64', '0x3fe40cea672819d5', '0x3fe1c83754e72c92', 2, 'np.float64', '0xbfa74770642e8ee0', '0xbfa74355fcf67332', 2, 'np.float64', '0x7fc60942d32c1285', '0x3ff0000000000000', 2)
```

which is correct

fast_executemany is a bit faster, not much.  script takes .8 seconds with fast_executemany and about 1 second with insertmanyvalues.

**Reply by fdcastel** [reply]
I'm getting similar performance results from @kmatt:

```
SQLAlchemy 1.4, fast_executemany=True: 3462.7 rows/s

SQLAlchemy 2.0: 241.7 rows/s
SQLAlchemy 2.0, fast_executemany=True: 375.2 rows/s
SQLAlchemy 2.0, fast_executemany=True, use_insertmanyvalues=False: 3467.8 rows/s
```

Using:
- Microsoft ODBC Driver 18 for SQL Server (18.2.1.1)
- sqlalchemy = "2.0.8"
- pyodbc = "4.0.35"
- pandas = "1.5.3"


https://github.com/sqlalchemy/sqlalchemy/issues/9435 should not have been closed. Surely Pandas `.to_sql()` code is horrible. But 
- the problem is still there;
- it isn't well documented; and 
- it will take someone by surprise when migrating from 1.4 to 2.0.

For now, if you are using Pandas, `fast_executemany=True, use_insertmanyvalues=False` is currently the only way to get the same performance from SQLAlchemy 1.4. Using the default settings will give you ~10x slower results.

**Reply by zzzeek** [reply]
the issue was closed because it cannot be reproduced.  please try my test case at https://github.com/sqlalchemy/sqlalchemy/discussions/9436#discussioncomment-5232411

**Reply by zzzeek** [reply]
also adding, if someday someone takes the time to give us a reproducing script, the issue will be fixed likely the same day.    but if people just use their workarounds and leave, then it will not be fixed.

**Reply by fdcastel** [reply]
@zzzeek Maybe the problem occurs only with large datasets?

Using the following test code:

```python
from timeit import default_timer as timer

import logging
import pandas as pd
import sqlalchemy as sa


def test_performance():
    uri = 'mssql+pyodbc://******:******@*****/test_db?driver=ODBC Driver 18 for SQL Server&TrustServerCertificate=YES'

    # engine = sa.create_engine(uri, fast_executemany=True, use_insertmanyvalues=False)
    engine = sa.create_engine(uri)

    with engine.connect() as target_connection:
        total_start_time = timer()

        entity_name = 'test_data'
        schema_name = 'test_schema'
        json_file = f'./data/{entity_name}.json'
        log = logging.getLogger(entity_name)

        insert_mode = 'replace'
        for data in pd.read_json(json_file, lines=True, chunksize=10000):
            rows = len(data)

            start_time = timer()
            data.to_sql(entity_name, target_connection, schema_name, if_exists=insert_mode, index=False, chunksize=10000)
            elapsed_seconds = timer() - start_time

            log.info('Pushed %d rows in %.1f seconds (%.1f rows/s).', rows, elapsed_seconds, rows / elapsed_seconds)

            insert_mode = 'append'

        total_elapsed_seconds = timer() - total_start_time
        logging.info('Finished in %.1f seconds.', total_elapsed_seconds)
```

I got the following times when using `sa.create_engine(uri)`

```
INFO     test_data:test_perf.py:35 Pushed 10000 rows in 26.2 seconds (381.4 rows/s).
INFO     test_data:test_perf.py:35 Pushed 10000 rows in 26.0 seconds (384.3 rows/s).
INFO     test_data:test_perf.py:35 Pushed 10000 rows in 25.6 seconds (390.6 rows/s).
INFO     test_data:test_perf.py:35 Pushed 10000 rows in 32.4 seconds (308.8 rows/s).
INFO     test_data:test_perf.py:35 Pushed 2570 rows in 15.6 seconds (165.1 rows/s).
INFO     root:test_perf.py:40 Finished in 126.4 seconds.
```

And the following when using `sa.create_engine(uri, fast_executemany=True, use_insertmanyvalues=False)`

```
INFO     test_data:test_perf.py:30 Pushed 10000 rows in 2.0 seconds (5052.2 rows/s).
INFO     test_data:test_perf.py:30 Pushed 10000 rows in 1.4 seconds (7038.3 rows/s).
INFO     test_data:test_perf.py:30 Pushed 10000 rows in 1.4 seconds (6955.7 rows/s).
INFO     test_data:test_perf.py:30 Pushed 10000 rows in 1.5 seconds (6872.8 rows/s).
INFO     test_data:test_perf.py:30 Pushed 2570 rows in 0.5 seconds (5444.2 rows/s).
INFO     root:test_perf.py:35 Finished in 7.3 seconds.
```

[test_data.zip](https://github.com/sqlalchemy/sqlalchemy/files/11130278/test_data.zip)


_Attached: test_data.json (15.5 MB) compressed with zip. Uncompress it into a `data/' subfolder._

_Edit: fixed elapsed times (wrong copy & paste in the first version)_

**Reply by gordthompson** [reply]
> Maybe the problem occurs only with large datasets?

DataFrames with lots of rows will (not surprisingly) have an effect, but so will DataFrames with lots of columns.

I used the first 5000 rows of your sample data to run the test mentioned in my answer here:

https://github.com/sqlalchemy/sqlalchemy/discussions/9436#discussioncomment-5498257

```
SQLAlchemy version 2.0.8
fast_executemany = False
use_insertmanyvalues = True
5,000 rows inserted in 62.7 seconds

SQLAlchemy version 2.0.8
fast_executemany = True
use_insertmanyvalues = False
5,000 rows inserted in 4.0 seconds
```

(Yes, my test machine is very slow.) So why the big difference?

SQL Server uses stored procedures (like `sp_prepexec`) to execute statements sent by the ODBC driver, and SQL Server stored procedures are limited to 2100 parameters. So for a "manyvalues" statement to insert 3 rows, like

```sql
INSERT INTO tbl (col1, col2) VALUES (?, ?), (?, ?), (?, ?)
```

we need to send 6 parameter values. The more columns there are, the fewer rows we can insert using a single "manyvalues" statement before hitting the 2100-parameter limit.

Your DataFrame has 28 columns, so each such statement can only insert a maximum of 2097 // 28 = 74 rows. And since fast_executemany=False means that each INSERT is sent separately, we are actually sending 67 individual INSERT statements for the first 4,958 rows, followed by one more INSERT statement for the final 42 rows.

By contrast, with fast_executemany=True and use_insertmanyvalues=False we are sending the single-row INSERT statement once (`sp_prepare`) followed by the data for all 5000 rows in a single binary "blob" (called an ODBC "parameter array"), which the server unpacks and then proceeds to (effectively) call `sp_execute` 5000 times server-side.

**Reply by zzzeek** [reply]
@gordthompson OK so that's a bug then.    the use_insertmanyvalues setting is not effective here, and if we are not using RETURNING we should perhaps reorganize the default strategy used, what do you think?

**Reply by zzzeek** [reply]
OK @fdcastel fast_executemany defaults to False in SQLAlchemy 1.4, so here we will maintain that default, but we will change it so that if fast_executemany is set to True, it will supersede insertmanyvalues for an INSERT statement that does not have RETURNING, which means it will behave identically as 1.4.   1.4 had no support for RETURNING with "executemany" and pandas does not use this (since it was not possible previously).

**Comment by gordthompson** [comment]
For SQL Server 2016 and later there is also this option, which uses OPENJSON instead of executemany():

https://gist.github.com/gordthompson/1fb0f1c3f5edbf6192e596de8350f205

In some cases it can be significantly faster than `fast_executemany=True`

**Reply by fdcastel** [reply]
I ran the same test I submitted before, albeit in another server (the first one was running SQL Server 2014, with no OPENJSON support).

With `method=None`:

```
INFO     test_data:test_perf.py:82 Pushed 10000 rows in 3.4 seconds (2919.6 rows/s).
INFO     test_data:test_perf.py:82 Pushed 10000 rows in 1.8 seconds (5604.5 rows/s).
INFO     test_data:test_perf.py:82 Pushed 10000 rows in 1.8 seconds (5607.0 rows/s).
INFO     test_data:test_perf.py:82 Pushed 10000 rows in 1.8 seconds (5660.4 rows/s).
INFO     test_data:test_perf.py:82 Pushed 2570 rows in 1.2 seconds (2167.6 rows/s).
INFO     root:test_perf.py:87 Finished in 10.5 seconds.
```

With `method=mssql_insert_json`:

```
INFO     test_data:test_perf.py:82 Pushed 10000 rows in 3.2 seconds (3107.8 rows/s).
INFO     test_data:test_perf.py:82 Pushed 10000 rows in 2.1 seconds (4829.7 rows/s).
INFO     test_data:test_perf.py:82 Pushed 10000 rows in 2.1 seconds (4837.9 rows/s).
INFO     test_data:test_perf.py:82 Pushed 10000 rows in 2.1 seconds (4836.3 rows/s).
INFO     test_data:test_perf.py:82 Pushed 2570 rows in 1.1 seconds (2244.9 rows/s).
```

In both cases, `fast_executemany=True, use_insertmanyvalues=False`.

**Reply by zzzeek** [reply]
both methods have the same limitation that we originally had problems with which is that it creates a single huge memory blob, as well as that setinputsizes() support for odd string collations goes out the window, so if the speed differences are negligible we can just stick with fast_executemany.

**Comment by gordthompson** [comment]
I suspect that at least some of the performance issues might relate to the pyodbc issue here:

https://github.com/mkleehammer/pyodbc/issues/741

TL;DR - `fast_executemany` performance can be severely degraded under certain circumstances if the first row of a DataFrame contains Null/NaN values.

With SQLA 1.4 (before `use_insertmanyvalues`), pyodbc with fast_executemany would pack each row of the DataFrame into its own "row" in the ODBC parameter array and the INSERT statement would look like

```sql
INSERT INTO tbl (col1, col2) VALUES (?, ?)
```

Under SQLA 2.0 with `use_insertmanyvalues=True`, we generate an INSERT statement like

```sql
INSERT INTO tbl (col1, col2) VALUES (?, ?), (?, ?), (?, ?)
```

and the first "row" of the parameter array actually contains the first three rows of the DataFrame. That makes it three times more likely that one or more Null/NaN values can slow pyodbc down.

The vexing part is that the performance degradation is highly variable: Sometimes it is quite modest and other times it is an order of magnitude or more. I did a quick informal test using sample data from the above-cited pyodbc issue and got results like this:

```
SQLAlchemy version 1.4.47
fast_executemany = False
1,998 rows inserted in 62.8 seconds

SQLAlchemy version 1.4.47
fast_executemany = True
1,998 rows inserted in 15.8 seconds

SQLAlchemy version 2.0.8
fast_executemany = False
use_insertmanyvalues = True
1,998 rows inserted in 16.8 seconds

SQLAlchemy version 2.0.8
fast_executemany = True
use_insertmanyvalues = True
1,998 rows inserted in 20.6 seconds

SQLAlchemy version 2.0.8
fast_executemany = True
use_insertmanyvalues = False
1,998 rows inserted in 13.7 seconds
```

The good news is that as I mentioned above, with SQL Server 2016 and newer we can use an alternate method that avoids `.executemany()` (and hence fast_executemany) altogether

https://gist.github.com/gordthompson/1fb0f1c3f5edbf6192e596de8350f205

and that same test gave me

```
SQLAlchemy version 2.0.8
fast_executemany = (n/a)
1,998 rows inserted in 2.9 seconds
```

**Reply by zzzeek** [reply]
@gordthompson that's yet another method of "insertmanyvalues" that would supersede what we have.

~~@gordthompson can you open a new issue where we can work out what the best strategies for SQL Server can be for the non-RETURNING case ?   as far as the JSON thing that seems a little bit out there for the moment.~~  I've opened #9586

**Reply by dsb1Beno** [reply]
Thank you @gordthompson for providing your custom mssql_insert_json function. I wasn't satisfied with the default to_sql performance, with fast_executemany enabled. 

I was able to adopt your mssql_insert_json function and trim another minute off my upload time. 
Moreover, I played with different chunk sizes and observed that using a value anywhere from 50k to 100k was optimal for my dataframes.

Here are some performance stats and some context.

Environment:
Microsoft Azure SQL Managed Instance: General Purpose Standard-series (Gen 5) (4 vCores, 20GB RAM)
Dialect: mssql+pyodbc
Driver: ODBC Driver 17 for SQL Server

Package version:
SQLAlchemy: 2.0.30
Pandas: 2.2.2
pyodbc: 5.0.1

Data: 
~5.3 million rows across 7 dataframes

Performance:
fast_executemany=True / no custom method / no chunksize (sent all at once)  -->  03:55
fast_executemany=True / no custom method / 100k chunksize  -->  03:39

fast_executemany=N/a / mssql_insert_json / 10k chunksize  -->  02:47
fast_executemany=N/a / mssql_insert_json / 25k chunksize  -->  02:28
fast_executemany=N/a / mssql_insert_json / 50k chunksize  -->  02:22
fast_executemany=N/a / mssql_insert_json / 100k chunksize  -->  02:22
fast_executemany=N/a / mssql_insert_json / 200k chunksize  -->  02:26


I tried implementing multi-threading with concurrent.futures.ThreadPoolExecutor to further boost performance, but I kept running into this deadlock error, which I think is just a MSSQL bottleneck:
sqlalchemy.exc.DBAPIError: (pyodbc.Error) ('40001', '[40001] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Transaction (Process ID 111) was deadlocked on lock resources with another process and has been chosen as the deadlock victim. Rerun the transaction. (1205) (SQLExecDirectW)')

Anyways, sorry to bump an old thread, simply wanted to provide an update to any other builders that want to boost performance when writing data to MSSQL from Pandas dataframes.
